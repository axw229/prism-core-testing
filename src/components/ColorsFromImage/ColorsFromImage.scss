.match-photo-picker {
  &__wrapper {
    width: 100%;
    /* @todo - This just stopped the annoy jank, needs to be properly fixed. -RS */
    cursor: pointer;
  }
  &__canvas {
    &--portrait {
      margin: 0 auto;
      text-align: center;
      display: block;
    }
  }
}