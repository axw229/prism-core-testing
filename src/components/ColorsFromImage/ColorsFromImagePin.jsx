// @flow
import React, { PureComponent, Fragment } from 'react'

type Props = {
  pinType: string,
  onClickMethod: Function,
  // activePinIndex: number,
  isActiveFlag: boolean,
  transformValue: string,
  RGBstring: string,
  previewColorName: string,
  previewColorNumber: string
}

class ColorsFromImagePin extends PureComponent<Props> {
  render () {
    // eslint-disable-next-line one-var
    const { previewColorName, previewColorNumber, transformValue, RGBstring, onClickMethod, isActiveFlag, pinType } = this.props

    let isPreviewPin = false
    let additionalPinClass = ''
    let isActiveClass = ''

    if (pinType === 'preview') {
      isPreviewPin = true
      additionalPinClass = 'pin--preview__wrapper'
    }

    if (isPreviewPin) {
      if (isActiveFlag) isActiveClass = 'pin--active__wrapper'
    } else {
      // TODO: fix this functionality
      // is `this.key` supposed to be a reference to the key prop? is it supposed to be internally tracked?
      // if (activePinIndex === this.key) isActiveClass = 'pin--active__wrapper'
    }

    return (
      <Fragment>
        <button onClick={onClickMethod} style={{ transform: transformValue }} className={additionalPinClass + ' pin__wrapper ' + isActiveClass}>
          <span className='pin__chip' style={{ background: RGBstring }} />
          <div className='pin__name-wrapper'>
            <span className='pin__copy pin__name'>{previewColorName}</span>
            <span className='pin__copy pin__number'>SW {previewColorNumber}</span>
          </div>
        </button>
      </Fragment>
    )
  }
}

export default ColorsFromImagePin
