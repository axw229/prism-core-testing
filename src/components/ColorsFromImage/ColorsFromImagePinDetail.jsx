<div class="color-detail__wrapper">
  <div class="color-detail">
    <div class="scene__room__tinted-wrapper">
      <img class="scene__room scene__room--tinted" src="./images/room.jpg" alt="" />
    </div>
    <img class="scene__room scene__room--natural" src="./images/room.jpg" alt="" />
  </div>
</div>