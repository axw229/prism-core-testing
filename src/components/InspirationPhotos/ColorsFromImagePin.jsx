// @flow
import React, { useState, useRef, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { useIntl } from 'react-intl'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { add } from '../../store/actions/live-palette'
import type { Color } from 'src/shared/types/Colors.js.flow'
import { LiveMessage } from 'react-aria-live'
import { activedPinsHalfWidth } from './data'
import { getContrastYIQ } from '../../../src/shared/helpers/ColorUtils'
import { KEY_CODES } from 'src/constants/globals'
import './ColorPins.scss'
import 'src/scss/convenience/visually-hidden.scss'
import 'src/providers/fontawesome/fontawesome'

type Props = {
  color: {
    ...Color,
    isActiveFlag: boolean,
    isContentLeft: boolean,
    translateX: number,
    translateY: number,
    rgbValue: string,
    pinNumber: number,
    isMovingPin: boolean,
    hide: boolean,
    pinId: string
  },
  useTranslateProps: boolean,
  translateX: number,
  translateY: number,
  activatePin: Function,
  handleDrag: Function,
  deleteCurrentPin: Function,
  handleDragStop: Function,
  handlePinMoveByKeyboard: Function,
  handleKeyUpAfterPinMove: Function,
  handleTouchEnd: Function
}

export const CLASSES = { BASE: 'pin__wrapper', CHECKBOX: 'visually-hidden' }

const PIN_MOVEMENT_INTERVAL = 10
const PIN_MOVEMENT_SHIFT_KEY_INTERVAL = 1

export default (props: Props) => {
  const { color, activatePin, handlePinMoveByKeyboard, handleKeyUpAfterPinMove, deleteCurrentPin, handleDrag, handleDragStop, handleTouchEnd } = props

  const dispatch = useDispatch()
  const { formatMessage } = useIntl()
  const livePaletteColors = useSelector(store => store.lp.colors)
  const colorDivRef: { current?: HTMLDivElement } = useRef()
  const [isColorDivFocused: boolean, setIsColorDivFocused: boolean => void] = useState()
  const isColorAdded = livePaletteColors.some(paletteColor => paletteColor.colorNumber === color.colorNumber)
  const [isHidden, setIsHidden] = useState(false)

  useEffect(() => {
    setIsHidden(color.hide)
  }, [color])

  const focusHandler = () => {
    colorDivRef.current && colorDivRef.current.focus()
    if (isColorDivFocused && colorDivRef.current) {
      colorDivRef.current.style.outlineStyle = 'solid'
      colorDivRef.current.style.outlineWidth = '2px'
      colorDivRef.current.style.outlineColor = '#2cabe2'
    }
  }

  const getElementsTopLeftCoordinates = (elem) => {
    const { top, left } = elem.target.getBoundingClientRect()
    return { top, left }
  }

  const onKeyDownHandlerDiv = (e) => {
    e.keyCode === KEY_CODES.KEY_CODE_TAB && setIsColorDivFocused(true)

    if (isColorDivFocused && (e.keyCode === KEY_CODES.KEY_CODE_ENTER || e.keyCode === KEY_CODES.KEY_CODE_SPACE)) {
      e.stopPropagation()
      e.preventDefault()
      activatePin(color.pinNumber)
      focusHandler()
    } else if (isColorDivFocused &&
      color.isActiveFlag && (
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_LEFT ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_UP ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_RIGHT ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_DOWN)
    ) {
      e.stopPropagation()
      e.preventDefault()
      const pinPositionDocument = getElementsTopLeftCoordinates(e)
      const movementInterval = e.shiftKey ? PIN_MOVEMENT_SHIFT_KEY_INTERVAL : PIN_MOVEMENT_INTERVAL
      let pinTopReferredToCanvas = Math.floor(pinPositionDocument.top) + activedPinsHalfWidth
      let pinLeftReferredToCanvas = Math.floor(pinPositionDocument.left) + activedPinsHalfWidth

      switch (e.keyCode) {
        case KEY_CODES.KEY_CODE_ARROW_LEFT:
          pinLeftReferredToCanvas -= movementInterval
          break
        case KEY_CODES.KEY_CODE_ARROW_UP:
          pinTopReferredToCanvas -= movementInterval
          break
        case KEY_CODES.KEY_CODE_ARROW_RIGHT:
          pinLeftReferredToCanvas += movementInterval
          break
        case KEY_CODES.KEY_CODE_ARROW_DOWN:
          pinTopReferredToCanvas += movementInterval
          break
      }

      handlePinMoveByKeyboard({ offsetX: pinLeftReferredToCanvas, offsetY: pinTopReferredToCanvas, pinNumber: color.pinNumber })
    }
  }

  const handleKeyUpDiv = (e) => {
    if (e.keyCode === KEY_CODES.KEY_CODE_ENTER || e.keyCode === KEY_CODES.KEY_CODE_SPACE) {
      focusHandler()
    } else if (color.isActiveFlag && (
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_LEFT ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_UP ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_RIGHT ||
      e.keyCode === KEY_CODES.KEY_CODE_ARROW_DOWN
    )) {
      handleKeyUpAfterPinMove(color.pinNumber)
    }
  }

  const handleMouseMove = (e: SyntheticEvent, isMobile: booleam = false) => {
    if (e.target) {
      e.stopPropagation()
    }

    handleDrag(e.clientX, e.clientY, isMobile)
  }

  const handlePinDragStop = (e: SyntheticEvent) => {
    e.stopPropagation()
    handleDragStop(e)
    window.removeEventListener('mousemove', handleMouseMove)
    window.removeEventListener('mouseup', handlePinDragStop)
  }

  const handlePinDragStart = (e) => {
    deleteCurrentPin(color.pinNumber)
    window.addEventListener('mousemove', handleMouseMove)
    window.addEventListener('mouseup', handlePinDragStop)
  }

  const handleTouchStart = (e: SyntheticEvent) => {
    setIsColorDivFocused(false)
    setIsHidden(true)
  }

  const handleTouchStop = (e: SyntheticEvent) => {
    const dims = colorDivRef.current.getBoundingClientRect()
    handleTouchEnd(dims, color.pinId)
  }

  const handleTouchMove = (e: SyntheticEvent) => {
    // @todo this gets drag working on mobile but color is off, this is index math issue.  I should build a translator...
    handleMouseMove({ clientX: e.targetTouches[0].clientX, clientY: e.targetTouches[0].clientY }, true)
  }

  const translateX = props.useTranslateProps ? props.translateX : color.translateX
  const translateY = props.useTranslateProps ? props.translateY : color.translateY
  return (
    <>
      <label
        title={color.name}
        aria-label={color.name}
        htmlFor={color.name}
        tabIndex='-1'
        onClick={(e) => {
          e.preventDefault()
          e.stopPropagation()
          activatePin(color.pinNumber)
        }}
        style={{ transform: `translate(${translateX - 360}px, ${translateY}px)`, overflow: 'visible', display: isHidden ? 'none' : 'flex' }}
        className={`pin__wrapper ${color.isActiveFlag ? 'pin__wrapper--active' : ''}`}
      >
        <div className='pin__content__wrapper'>
          <div
            ref={colorDivRef}
            role='button'
            draggable
            tabIndex='0'
            onKeyDown={onKeyDownHandlerDiv}
            onKeyUp={handleKeyUpDiv}
            onFocus={focusHandler}
            onBlur={() => {
              colorDivRef.current.blur()
              colorDivRef.current.style.outline = 'none'
            }}
            onMouseDown={() => setIsColorDivFocused(false)}
            onDragStart={handlePinDragStart}
            onTouchStart={handleTouchStart}
            onTouchMove={handleTouchMove}
            onTouchEnd={handleTouchStop}
            className={`pin__chip ${(color.isActiveFlag) ? `pin__chip--active` : ''}`}
            style={{ background: color.rgbValue }}
          >
            <input
              tabIndex='-1'
              className={CLASSES.CHECKBOX}
              type='checkbox'
              checked={color.isActiveFlag}
              name={`pin-${color.pinNumber}`}
              id={`pin-${color.pinNumber}`}
              onChange={() => activatePin(color.pinNumber)}
            />
            {(isColorAdded && !color.isActiveFlag) && (<FontAwesomeIcon style={{ color: getContrastYIQ(color.hex) }} icon={['fa', 'check-circle']} size='1x' />)}
          </div>
          <div className={`pin__content ${color.isContentLeft ? 'pin__content--left' : ''}`}>
            <div className={`pin__name__wrapper ${color.isActiveFlag ? `pin__name__wrapper--active` : `pin__name__wrapper--inactive`}`}>
              <span className='pin__copy pin__number'>{(color.colorNumber) ? `${color.brandKey}${color.colorNumber}` : ''}</span>
              <span className='pin__copy pin__name'>{color.name}</span>
            </div>
            <div
              className={`pin__wrapper__pin-add ${color.isContentLeft ? 'pin__wrapper__pin-add--left' : ''} ${color.isActiveFlag ? `pin__wrapper__pin-add--active` : ''}`}
              style={{ background: color.rgbValue }}
            >
              <div className={`pin__wrapper__pin-add-icon ${color.isContentLeft ? 'pin__wrapper__pin-add-icon--left' : ''}`}>
                {(isColorAdded)
                  ? (<FontAwesomeIcon style={{ color: getContrastYIQ(color.hex) }} icon={['fa', 'check-circle']} size='2x' />)
                  : (
                    <button
                      tabIndex={(color.isActiveFlag) ? `0` : `-1`}
                      className={`pin__wrapper__pin-add-button`}
                      onClick={(e) => {
                        e.stopPropagation()
                        dispatch(add(color))
                      }}
                    >
                      <FontAwesomeIcon tabIndex='-1' style={{ color: getContrastYIQ(color.hex) }} icon={['fal', 'plus-circle']} size='2x' />
                    </button>
                  )
                }
              </div>
            </div>
          </div>
        </div>
      </label>
      {color.isActiveFlag && <LiveMessage message={formatMessage({ id: 'LIVE_MESSAGE_COLOR_NAME' }, { previewColorName: color.colorName })} aria-live='polite' />}
      {color.isActiveFlag && isColorDivFocused && <LiveMessage message={formatMessage({ id: 'LIVE_MESSAGE_MOVE_PIN' })} aria-live='assertive' clearOnUnmount='true' />}
    </>
  )
}
