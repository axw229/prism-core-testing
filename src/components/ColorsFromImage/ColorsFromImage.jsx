'use strict'

import React, { PureComponent, Fragment, createRef } from 'react'
import { brandColors } from './sw-colors-in-LAB.js'
import { getDeltaE00 } from 'delta-e'
import ColorsFromImagePin from './ColorsFromImagePin'

import './ColorsFromImage.scss'

const baseClass = 'match-photo-picker'
const wrapperClass = `${baseClass}__wrapper`
const canvasClass = `${baseClass}__canvas`
const hiddenImageClass = `${baseClass}__hidden-image`
const portraitOrientation = `${canvasClass}--portrait`

class ColorsFromImage extends PureComponent {
  constructor (props) {
    super(props)

    this.state = {
      previewPinIsUpdating: false,
      previewPinIsActive: false,
      previewColorName: '',
      previewColorNumber: '',
      cursorIsPaused: false,
      cursorX: 0,
      cursorY: 0,
      previewPinX: 0,
      previewPinY: 0,
      mappedCanvasIndex: 0,
      currentPixelRGB: [0, 0, 0],
      currentPixelRGBstring: 'rgb(0,0,0)',
      currentBrandColorIndex: null,
      pinnedColors: [],
      activePinIndex: -1,
      currentPinIndex: -1,
      imageStatus: 'loading'
    }

    this.findBrandColorTimeout = null
    this.brandColorsLength = brandColors.length
    this.CFICanvas = createRef()
    this.CFIWrapper = createRef()
    this.imageRef = createRef()
    this.handleImageLoaded = this.handleImageLoaded.bind(this)
    this.handleImageErrored = this.handleImageErrored.bind(this)
    this.initCanvas = this.initCanvas.bind(this)
    this.wrapperMouseMove = this.wrapperMouseMove.bind(this)
    this.rgb2lab = this.rgb2lab.bind(this)
    this.findBrandColor = this.findBrandColor.bind(this)
    this.updatePreviewPin = this.updatePreviewPin.bind(this)
  }

  rgb2lab (rgb) {
    // eslint-disable-next-line one-var
    var r = rgb[0] / 255,
      g = rgb[1] / 255,
      b = rgb[2] / 255,
      x, y, z

    r = (r > 0.04045) ? Math.pow((r + 0.055) / 1.055, 2.4) : r / 12.92
    g = (g > 0.04045) ? Math.pow((g + 0.055) / 1.055, 2.4) : g / 12.92
    b = (b > 0.04045) ? Math.pow((b + 0.055) / 1.055, 2.4) : b / 12.92

    x = (r * 0.4124 + g * 0.3576 + b * 0.1805) / 0.95047
    y = (r * 0.2126 + g * 0.7152 + b * 0.0722) / 1.00000
    z = (r * 0.0193 + g * 0.1192 + b * 0.9505) / 1.08883

    x = (x > 0.008856) ? Math.pow(x, 1 / 3) : (7.787 * x) + 16 / 116
    y = (y > 0.008856) ? Math.pow(y, 1 / 3) : (7.787 * y) + 16 / 116
    z = (z > 0.008856) ? Math.pow(z, 1 / 3) : (7.787 * z) + 16 / 116

    return [(116 * y) - 16, 500 * (x - y), 200 * (y - z)]
  }

  handleImageLoaded () {
    this.setState({ imageStatus: 'loaded' })
    this.initCanvas()
  }

  handleImageErrored () {
    this.setState({ imageStatus: 'failed' })
  }

  initCanvas () {
    this.CFICanvasContext = this.CFICanvas.current.getContext('2d')
    // eslint-disable-next-line react/prop-types
    this.handleResize(this.props.width, this.props.height)
  }

  addPin () {
    console.log('Adding pin')
  }

  handleResize (width, height) {
    // @todo - determine if I need to set/get the pixeldata
    this.CFICanvasContext.clearRect(0, 0, width, height)
    // eslint-disable-next-line react/prop-types
    this.CFICanvasContext.drawImage(this.imageRef.current, 0, 0, width, height)

    this.canvasOffset = this.CFICanvas.current.getBoundingClientRect()
    this.wrapperOffset = this.CFIWrapper.current.getBoundingClientRect()
  }

  findBrandColor () {
    let currentPixelInLABarray = this.rgb2lab(this.state.currentPixelRGB)
    let currentPixelInLAB = { L: currentPixelInLABarray[0], A: currentPixelInLABarray[1], B: currentPixelInLABarray[2] }

    for (var arrayIndex = 0; arrayIndex < this.brandColorsLength; arrayIndex += 6) {
      let thisSWinLAB = { L: brandColors[arrayIndex + 3], A: brandColors[arrayIndex + 4], B: brandColors[arrayIndex + 5] }
      let colorDistance = getDeltaE00(currentPixelInLAB, thisSWinLAB)

      if (colorDistance < 12) {
        this.setState({ currentBrandColorIndex: arrayIndex, currentPixelRGB: brandColors[arrayIndex + 2], currentPixelRGBstring: 'rgb(' + brandColors[arrayIndex + 2] + ')', previewPinIsActive: true, previewColorName: brandColors[arrayIndex], previewColorNumber: brandColors[arrayIndex + 1] })
        arrayIndex = this.brandColorsLength // This is to kick us out of the inner loop, since we've found a good match
      }
    }
  }

  updatePreviewPin () {
    let _this = this
    this.previewPinIsUpdating = true

    // This is manually debouncing cursor position updates
    window.setTimeout(function () { _this.setState({ previewPinIsUpdating: false }) }, 50)

    // Clear the previous mouse move call to findBrandColor and set a new one for the current cursor position
    window.clearTimeout(this.findBrandColorTimeout)
    this.findBrandColorTimeout = window.setTimeout(this.findBrandColor, 700)
  }

  // @todo - fix coords -RS
  wrapperMouseMove (e) {
    if (!this.state.previewPinIsUpdating) {
      const cursorX = e.clientX - this.wrapperOffset.left // can also try screenX or pageX
      const cursorY = e.clientY - this.wrapperOffset.top
      // const mappedCanvasIndex = (cursorY * this.state.canvasWidth + cursorX) * 4
      // Using a bitwise or to trunc here since it is much faster than something like Math.trunc()
      const mappedCanvasIndex = (((e.clientY - this.canvasOffset.top) * this.state.canvasWidth) + (e.clientX - this.canvasOffset.left | 0)) * 4
      // eslint-disable-next-line react/prop-types
      const imageData = this.props.data.imageData.data

      this.setState({
        currentPixelRGB: [imageData[mappedCanvasIndex], imageData[mappedCanvasIndex + 1], imageData[mappedCanvasIndex + 2]],
        currentPixelRGBstring: 'rgb(' + imageData[mappedCanvasIndex] + ',' + imageData[mappedCanvasIndex + 1] + ',' + imageData[mappedCanvasIndex + 2] + ')',
        mappedCanvasIndex: mappedCanvasIndex,
        previewPinIsUpdating: true,
        previewPinIsActive: false,
        previewPinX: cursorX,
        previewPinY: cursorY }, () => {
        (() => { this.updatePreviewPin() })()
      })
    }
  }

  componentDidMount () {
    this.initCanvas()
  }

  componentDidUpdate (prevProps, prevState) {
    // eslint-disable-next-line react/prop-types
    if (prevProps.width !== this.props.width) {
      // eslint-disable-next-line react/prop-types
      this.handleResize(prevProps.width, prevProps.height)
    }
  }

  render () {
    // eslint-disable-next-line react/prop-types
    const { isPortrait, img } = this.props.data

    const { previewPinIsActive, previewPinX, previewPinY, previewColorName, previewColorNumber, currentPixelRGBstring } = this.state
    const transformValue = `translate(${previewPinX}px, ${previewPinY}px)`

    return (
      <Fragment>
        <img ref={this.imageRef} className={hiddenImageClass} src={img} onLoad={this.handleImageLoaded} alt='hidden' />
        <div onMouseMove={this.wrapperMouseMove} className={wrapperClass} ref={this.CFIWrapper}>
          {/* eslint-disable-next-line react/prop-types */}
          <canvas className={isPortrait ? portraitOrientation : canvasClass} ref={this.CFICanvas} width={this.props.width} height={this.props.height} />
          <ColorsFromImagePin key='1'
            isActiveFlag={previewPinIsActive}
            pinType='preview'
            previewColorName={previewColorName}
            previewColorNumber={previewColorNumber}
            RGBstring={currentPixelRGBstring}
            transformValue={transformValue}
            onClickMethod={this.addPin} />
        </div>
      </Fragment>
    )
  }
}

export default ColorsFromImage

/*
-------------------------------------------------------------------------------------------
------------------ ### TO DO ### ------------------
-------------------------------------------------------------------------------------------
Do three steps of increasing fidelity checking in findBrandColor

Memoize result of findBrandColor

Confirm that I'm manually debouncing mousemove
*/
