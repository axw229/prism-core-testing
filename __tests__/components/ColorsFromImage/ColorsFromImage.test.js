import React from 'react'
import { mount } from 'enzyme'
import ColorsFromImage from 'src/components/ColorsFromImage/ColorsFromImage'
import ColorsFromImagePin from 'src/components/ColorsFromImage/ColorsFromImagePin'
import { brandColors } from 'src/components/ColorsFromImage/sw-colors-in-LAB.js'

const createColorsFromImage = (props) => {
  return mount(<ColorsFromImage {...props} />)
}

describe('snapshot testing', () => {
  it('should match snaphshot', () => {
    const wrapper = createColorsFromImage()
    expect(wrapper).toMatchSnapshot()
  })
})

describe('color from image compoment rendring test', () => {
  const wrapper = createColorsFromImage()
  it('should rendering colorFromImagePin Component', () => {
    expect(wrapper.find(ColorsFromImagePin).exists()).toBe(true)
  })

  it('should rendering canvas correctly', () => {
    expect(wrapper.find('canvas').exists()).toBe(true)
  })
})

// canvas testing
describe('canvas testing', () => {
  const wrapper = createColorsFromImage()
  it('event mousemove should be called when mouse move', () => {
    const canvasWrapper = wrapper.find('div').at(1)
    const spy = jest.fn()
    wrapper.instance().updatePreviewPin = spy
    canvasWrapper.simulate('mousemove')
    wrapper.instance().forceUpdate()
    expect(spy).toHaveBeenCalled()
  })

  it('should update state correctly', () => {
    const canvas = wrapper.instance()
    expect(canvas.state.imageStatus).toEqual('loading')
    canvas.handleImageLoaded()
    expect(canvas.state.imageStatus).toEqual('loaded')
    canvas.handleImageErrored()
    expect(canvas.state.imageStatus).toEqual('failed')
  })

  it('should rendring canvas correctly', () => {
    const canvas = wrapper.instance()
    const R = Math.floor(Math.random() * 255 + 1)
    const G = Math.floor(Math.random() * 255 + 1)
    const B = Math.floor(Math.random() * 255 + 1)
    const color = [R, G, B]

    expect(canvas.brandColorsLength).toEqual(brandColors.length)
    expect(canvas.canvasHeight).toEqual(487)
    expect(canvas.canvasWidth).toEqual(898)
    expect(Array.isArray(canvas.rgb2lab(color))).toBe(true)
  })
})
