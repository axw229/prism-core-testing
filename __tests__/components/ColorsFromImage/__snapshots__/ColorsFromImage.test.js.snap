// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing should match snaphshot 1`] = `
<ColorsFromImage>
  <div
    className="CFIwrapper "
    onMouseMove={[Function]}
  >
    <canvas
      className="CFIcanvas"
    />
    <img
      alt=""
      className="CFIimage"
      onError={[Function]}
      onLoad={[Function]}
      src="/prism/images/colors-from-image/frog.png"
    />
    <ColorsFromImagePin
      RGBstring="rgb(0,0,0)"
      isActiveFlag={false}
      key="1"
      onClickMethod={[Function]}
      pinType="preview"
      previewColorName=""
      previewColorNumber=""
      transformValue="translate(0px, 0px)"
    >
      <button
        className="pin--preview__wrapper pin__wrapper "
        onClick={[Function]}
        style={
          Object {
            "transform": "translate(0px, 0px)",
          }
        }
      >
        <span
          className="pin__chip"
          style={
            Object {
              "background": "rgb(0,0,0)",
            }
          }
        />
        <div
          className="pin__name-wrapper"
        >
          <span
            className="pin__copy pin__name"
          />
          <span
            className="pin__copy pin__number"
          >
            SW 
          </span>
        </div>
      </button>
    </ColorsFromImagePin>
  </div>
</ColorsFromImage>
`;
